commit/merkleroot/outcome.go

```diff
@@ -20,40 +19,47 @@ import (
 // - chooses the seq num ranges for the next round
 // - builds a report
 // - checks for the transmission of a previous report
-func (p *Plugin) Outcome(
-	outCtx ocr3types.OutcomeContext, query types.Query, aos []types.AttributedObservation,
-) (ocr3types.Outcome, error) {
+func (w *Processor) Outcome(
+	prevOutcome Outcome,
+	query Query,
+	aos []shared.AttributedObservation[Observation],
+) (Outcome, error) {
 	tStart := time.Now()
-	outcome, nextState := p.getOutcome(outCtx, aos)
-	p.lggr.Infow("Sending Outcome",
-		"outcome", outcome, "oid", p.nodeID, "nextState", nextState, "outcomeDuration", time.Since(tStart))
-	return outcome.Encode()
+	outcome, nextState := w.getOutcome(prevOutcome, query, aos)
+	w.lggr.Infow("Sending Outcome",
+		"outcome", outcome, "oid", w.nodeID, "nextState", nextState, "outcomeDuration", time.Since(tStart))
+	return outcome, nil
 }
 
-func (p *Plugin) getOutcome(
-	outCtx ocr3types.OutcomeContext, aos []types.AttributedObservation,
+func (w *Processor) getOutcome(
+	previousOutcome Outcome,
+	query Query,
+	aos []shared.AttributedObservation[Observation],
 ) (Outcome, State) {
-	previousOutcome, nextState := p.decodeOutcome(outCtx.PreviousOutcome)
-	commitQuery := Query{}
+	nextState := previousOutcome.NextState()
 
-	consensusObservation, err := getConsensusObservation(p.lggr, p.reportingCfg.F, p.cfg.DestChain, aos)
+	consensusObservation, err := getConsensusObservation(w.lggr, w.reportingCfg.F, w.cfg.DestChain, aos)
 	if err != nil {
-		p.lggr.Warnw("Get consensus observation failed, empty outcome", "error", err)
+		w.lggr.Warnw("Get consensus observation failed, empty outcome", "error", err)
 		return Outcome{}, nextState
 	}
 
+	var outcome Outcome
 	switch nextState {
 	case SelectingRangesForReport:
-		return reportRangesOutcome(commitQuery, consensusObservation), nextState
+		outcome = reportRangesOutcome(query, consensusObservation)
 	case BuildingReport:
-		return buildReport(commitQuery, consensusObservation, previousOutcome), nextState
+		outcome = buildReport(query, consensusObservation, previousOutcome)
 	case WaitingForReportTransmission:
-		return checkForReportTransmission(
-			p.lggr, p.cfg.MaxReportTransmissionCheckAttempts, previousOutcome, consensusObservation), nextState
+		outcome = checkForReportTransmission(
+			w.lggr, w.cfg.MaxReportTransmissionCheckAttempts, previousOutcome, consensusObservation)
 	default:
-		p.lggr.Warnw("Unexpected next state in Outcome", "state", nextState)
+		w.lggr.Warnw("Unexpected next state in Outcome", "state", nextState)
 		return Outcome{}, nextState
 	}
+
+	outcome.Sort()
```

Comment:
Why is this refactoring necessary?

Reply:
We need to make sure the lists in the outcome are sorted to be deterministic. The sort was being used in the Encoding step for the Outcome, I moved it inside the Outcome method itself so that the Encode just encodes whatever comes.